cdc/capture/http_handler_test.go

```diff
@@ -0,0 +1,441 @@
+			changeFeedID + "2": {State: model.StateStopped},
+		}, nil)
+
+	statusProvider.On("GetAllTaskStatuses", mock.Anything).
+		Return(map[model.CaptureID]*model.TaskStatus{captureID: {}}, nil)
+
+	statusProvider.On("GetChangeFeedInfo", mock.Anything).
+		Return(&model.ChangeFeedInfo{State: model.StateNormal}, nil)
+
+	statusProvider.On("GetProcessors", mock.Anything).
+		Return([]*model.ProcInfoSnap{{CfID: changeFeedID, CaptureID: captureID}}, nil)
+
+	statusProvider.On("GetCaptures", mock.Anything).
+		Return([]*model.CaptureInfo{{ID: captureID}}, nil)
+
+	return statusProvider
+}
+
+func TestListChangefeed(t *testing.T) {
+	t.Parallel()
+	router := newRouter(newStatusServer())
+
+	// test list changefeed succeeded
+	api := openAPI{url: "/api/v1/changefeeds", method: "GET"}
+	w := httptest.NewRecorder()
+	req, _ := http.NewRequest(api.method, api.url, nil)
+	router.ServeHTTP(w, req)
+	require.Equal(t, 200, w.Code)
+	var resp []model.ChangefeedCommonInfo
+	err := json.NewDecoder(w.Body).Decode(&resp)
+	require.Nil(t, err)
+	require.Equal(t, 2, len(resp))
+
+	// test list changefeed with specific state
+	api = openAPI{url: fmt.Sprintf("/api/v1/changefeeds?state=%s", "stopped"), method: "GET"}
+	w = httptest.NewRecorder()
+	req, _ = http.NewRequest(api.method, api.url, nil)
+	router.ServeHTTP(w, req)
+	require.Equal(t, 200, w.Code)
+	resp = []model.ChangefeedCommonInfo{}
+	err = json.NewDecoder(w.Body).Decode(&resp)
+	require.Nil(t, err)
+	require.Equal(t, 1, len(resp))
+	require.Equal(t, model.StateStopped, resp[0].FeedState)
+	require.Equal(t, uint64(0x2), resp[0].CheckpointTSO)
+}
+
+func TestGetChangefeed(t *testing.T) {
+	t.Parallel()
+	router := newRouter(newStatusServer())
+
+	// test get changefeed succeeded
+	api := openAPI{url: fmt.Sprintf("/api/v1/changefeeds/%s", changeFeedID), method: "GET"}
+	w := httptest.NewRecorder()
+	req, _ := http.NewRequest(api.method, api.url, nil)
+	router.ServeHTTP(w, req)
+	require.Equal(t, 200, w.Code)
+	var resp model.ChangefeedDetail
+	err := json.NewDecoder(w.Body).Decode(&resp)
+	require.Nil(t, err)
+	require.Equal(t, model.StateNormal, resp.FeedState)
+
+	// test get changefeed failed
+	api = openAPI{url: fmt.Sprintf("/api/v1/changefeeds/%s", nonExistChanfeedID), method: "GET"}
+	w = httptest.NewRecorder()
+	req, _ = http.NewRequest(api.method, api.url, nil)
+	router.ServeHTTP(w, req)
+	require.Equal(t, 400, w.Code)
+	respErr := model.HTTPError{}
+	err = json.NewDecoder(w.Body).Decode(&respErr)
+	require.Nil(t, err)
+	require.Contains(t, respErr.Error, "changefeed not exists")
+}
+
+func TestPauseChangefeed(t *testing.T) {
+	t.Parallel()
+	router := newRouter(newStatusServer())
+	// test pause changefeed succeeded
+	api := openAPI{url: fmt.Sprintf("/api/v1/changefeeds/%s/pause", changeFeedID), method: "POST"}
+	w := httptest.NewRecorder()
+	req, _ := http.NewRequest(api.method, api.url, nil)
+	router.ServeHTTP(w, req)
+	require.Equal(t, 202, w.Code)
+
```

Comment:
Is it necessary to add test cases for pausing/resuming/removing changefeed twice?

Reply:
Since the unit tests here will not test the logic inside the owner, we can’t tell from the unit test whether these APIs are idempotent, so I don’t think we need to add these test cases.
We can do such test in HTTP API integration tests.